shopfloor_reception/services/reception.py

```diff
@@ -275,45 +326,165 @@ def _scan_document__by_picking(self, barcode):
                 )
             return self._select_picking(reception_pickings)
 
-    def _scan_document__by_product(self, barcode):
-        search = self._actions_for("search")
-        product = search.product_from_scan(barcode)
+    def _scan_document__by_product(self, product):
         if product:
             return self._select_document_from_product(product)
 
-    def _scan_document__by_packaging(self, barcode):
-        search = self._actions_for("search")
-        packaging = search.packaging_from_scan(barcode)
+    def _scan_document__by_packaging(self, packaging):
         if packaging:
             return self._select_document_from_packaging(packaging)
 
-    def _scan_line__by_product(self, picking, barcode):
-        search = self._actions_for("search")
-        product = search.product_from_scan(barcode)
-        if product:
-            move = picking.move_lines.filtered(lambda m: m.product_id == product)
-            message = self._check_move_available(move, "product")
-            if message:
-                return self._response_for_select_move(
-                    picking,
-                    message=message,
-                )
-            return self._select_line_from_product(picking, move, product)
+    def _scan_document__by_delivered_picking(self, pickings):
+        if self.work.menu.allow_rma:
+            outgoing_pickings = pickings.filtered(
+                lambda p: (p.picking_type_code == "outgoing")
+            )
+            # could there be more than 1?
+            # TODO test with multi step pickings
+            outgoing_picking = fields.first(outgoing_pickings)
+            if outgoing_picking:
+                return_picking = self._scan_document__create_return(outgoing_picking)
+                return self._response_for_select_move(return_picking)
+        # A return picking has been scanned, but allow rma is disabled.
+        return self._scan_document__fallback(self.env["stock.picking"])
+
+    def _scan_document__fallback(self, empty_recordset):
+        assert not empty_recordset, "recordset in fallback handler should be empty"
+        return self._response_for_select_document(
+            message=self.msg_store.barcode_not_found()
+        )
 
-    def _scan_line__by_packaging(self, picking, barcode):
-        search = self._actions_for("search")
-        packaging = search.packaging_from_scan(barcode)
-        if packaging:
-            move = picking.move_lines.filtered(
-                lambda m: packaging in m.product_id.packaging_ids
+    def _scan_line__create_return_move(self, return_picking, origin_move):
+        # copied from odoo/src/addons/stock/wizard/stock_picking_return.py
+        product = origin_move.product_id
```

Comment:
the whole method split in smaller methods could go to shopfloor/actions/stock IMO